// remult/remult-sql-js.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SqlJsDataProvider = void 0;
var tslib_1 = require("tslib");
var sql_database_js_1 = require("./src/data-providers/sql-database.js");
var filter_consumer_bridge_to_sql_request_js_1 = require("./src/filter/filter-consumer-bridge-to-sql-request.js");
var RepositoryImplementation_js_1 = require("./src/remult3/RepositoryImplementation.js");
var SqlJsDataProvider = /** @class */ (function () {
    function SqlJsDataProvider(db) {
        this.db = db;
    }
    SqlJsDataProvider.prototype.getLimitSqlSyntax = function (limit, offset) {
        return ' limit ' + limit + ' offset ' + offset;
    };
    SqlJsDataProvider.prototype.createCommand = function () {
        return new SqlJsCommand(this.db);
    };
    SqlJsDataProvider.prototype.transaction = function (action) {
        throw new Error('Method not implemented.');
    };
    SqlJsDataProvider.prototype.entityIsUsedForTheFirstTime = function (entity) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.createTable(entity)];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    SqlJsDataProvider.prototype.ensureSchema = function (entities) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var entities_1, entities_1_1, entity, e_1_1;
            var e_1, _a;
            return tslib_1.__generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        _b.trys.push([0, 5, 6, 7]);
                        entities_1 = tslib_1.__values(entities), entities_1_1 = entities_1.next();
                        _b.label = 1;
                    case 1:
                        if (!!entities_1_1.done) return [3 /*break*/, 4];
                        entity = entities_1_1.value;
                        return [4 /*yield*/, this.createTable(entity)];
                    case 2:
                        _b.sent();
                        _b.label = 3;
                    case 3:
                        entities_1_1 = entities_1.next();
                        return [3 /*break*/, 1];
                    case 4: return [3 /*break*/, 7];
                    case 5:
                        e_1_1 = _b.sent();
                        e_1 = { error: e_1_1 };
                        return [3 /*break*/, 7];
                    case 6:
                        try {
                            if (entities_1_1 && !entities_1_1.done && (_a = entities_1.return)) _a.call(entities_1);
                        }
                        finally { if (e_1) throw e_1.error; }
                        return [7 /*endfinally*/];
                    case 7: return [2 /*return*/];
                }
            });
        });
    };
    SqlJsDataProvider.prototype.dropTable = function (entity) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var e, sql;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, (0, filter_consumer_bridge_to_sql_request_js_1.dbNamesOf)(entity)];
                    case 1:
                        e = _a.sent();
                        sql = 'drop  table if exists ' + e.$entityName;
                        if (sql_database_js_1.SqlDatabase.LogToConsole)
                            console.info(sql);
                        return [4 /*yield*/, this.createCommand().execute(sql)];
                    case 2:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    SqlJsDataProvider.prototype.addColumnSqlSyntax = function (x, dbName) {
        var result = dbName;
        var nullNumber = x.allowNull ? '' : ' default 0 not null';
        if (x.valueType == Date)
            result += ' integer';
        else if (x.valueType == Boolean)
            result += ' integer ' + nullNumber;
        else if (x.valueType == Number) {
            if (!x.valueConverter.fieldTypeInDb)
                result += ' real ' + nullNumber;
            else
                result += ' ' + x.valueConverter.fieldTypeInDb + ' ' + nullNumber;
        }
        else
            result += ' text' + (x.allowNull ? ' ' : " default '' not null ");
        return result;
    };
    SqlJsDataProvider.prototype.createTable = function (entity) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var result, e, _a, _b, x, sql;
            var e_2, _c;
            return tslib_1.__generator(this, function (_d) {
                switch (_d.label) {
                    case 0:
                        result = '';
                        return [4 /*yield*/, (0, filter_consumer_bridge_to_sql_request_js_1.dbNamesOf)(entity)];
                    case 1:
                        e = _d.sent();
                        try {
                            for (_a = tslib_1.__values(entity.fields), _b = _a.next(); !_b.done; _b = _a.next()) {
                                x = _b.value;
                                if (!(0, filter_consumer_bridge_to_sql_request_js_1.isDbReadonly)(x, e) || (0, RepositoryImplementation_js_1.isAutoIncrement)(x)) {
                                    if (result.length != 0)
                                        result += ',';
                                    result += '\r\n  ';
                                    if ((0, RepositoryImplementation_js_1.isAutoIncrement)(x)) {
                                        if (x.key != entity.idMetadata.field.key)
                                            throw 'in web sql, autoincrement is only allowed for primary key';
                                        result += e.$dbNameOf(x) + ' integer primary key autoincrement';
                                    }
                                    else {
                                        result += this.addColumnSqlSyntax(x, e.$dbNameOf(x));
                                        if (x.key == entity.idMetadata.field.key) {
                                            result += ' primary key';
                                        }
                                    }
                                }
                            }
                        }
                        catch (e_2_1) { e_2 = { error: e_2_1 }; }
                        finally {
                            try {
                                if (_b && !_b.done && (_c = _a.return)) _c.call(_a);
                            }
                            finally { if (e_2) throw e_2.error; }
                        }
                        sql = 'create table if not exists ' + e.$entityName + ' (' + result + '\r\n)';
                        if (sql_database_js_1.SqlDatabase.LogToConsole)
                            console.log(sql);
                        return [4 /*yield*/, this.createCommand().execute(sql)];
                    case 2:
                        _d.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    SqlJsDataProvider.prototype.wrapIdentifier = function (name) {
        return name;
    };
    return SqlJsDataProvider;
}());
exports.SqlJsDataProvider = SqlJsDataProvider;
var SqlJsCommand = /** @class */ (function () {
    function SqlJsCommand(db) {
        this.db = db;
        this.values = {};
        this.i = 0;
    }
    SqlJsCommand.prototype.execute = function (sql) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var _a, _b;
            return tslib_1.__generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        if (!(this.i == 0)) return [3 /*break*/, 2];
                        _a = SqlJsSqlResult.bind;
                        return [4 /*yield*/, this.db];
                    case 1: return [2 /*return*/, new (_a.apply(SqlJsSqlResult, [void 0, (_c.sent()).exec(sql)]))()];
                    case 2:
                        _b = SqlJsSqlResult.bind;
                        return [4 /*yield*/, this.db];
                    case 3: return [2 /*return*/, new (_b.apply(SqlJsSqlResult, [void 0, (_c.sent()).exec(sql, this.values)]))()];
                }
            });
        });
    };
    SqlJsCommand.prototype.addParameterAndReturnSqlToken = function (val) {
        if (val instanceof Date)
            val = val.valueOf();
        var key = ':' + ++this.i;
        this.values[key] = val;
        return key;
    };
    return SqlJsCommand;
}());
var SqlJsSqlResult = /** @class */ (function () {
    function SqlJsSqlResult(result) {
        var _a, _b;
        this.result = result;
        this.rows =
            (_b = (_a = result[0]) === null || _a === void 0 ? void 0 : _a.values.map(function (row) {
                return row.reduce(function (prev, curr, i) {
                    var _a;
                    return (tslib_1.__assign(tslib_1.__assign({}, prev), (_a = {}, _a[result[0].columns[i]] = curr, _a)));
                }, {});
            })) !== null && _b !== void 0 ? _b : [];
    }
    SqlJsSqlResult.prototype.getColumnKeyInResultForIndexInSelect = function (index) {
        var _a;
        return (_a = this.result[0]) === null || _a === void 0 ? void 0 : _a.columns[index];
    };
    return SqlJsSqlResult;
}());
