// remult/src/data-api.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.serializeError = exports.determineSort = exports.DataApi = void 0;
var tslib_1 = require("tslib");
var context_js_1 = require("./context.js");
var rest_data_provider_js_1 = require("./data-providers/rest-data-provider.js");
var filter_interfaces_js_1 = require("./filter/filter-interfaces.js");
var relationInfoMember_js_1 = require("./remult3/relationInfoMember.js");
var server_action_js_1 = require("./server-action.js");
var DataApi = /** @class */ (function () {
    function DataApi(repository, remult) {
        this.repository = repository;
        this.remult = remult;
    }
    DataApi.prototype.httpGet = function (res, req, serializeContext) {
        var action = req === null || req === void 0 ? void 0 : req.get('__action');
        if (action === null || action === void 0 ? void 0 : action.startsWith(rest_data_provider_js_1.liveQueryAction))
            return this.liveQuery(res, req, undefined, serializeContext, action.substring(rest_data_provider_js_1.liveQueryAction.length));
        switch (action) {
            case 'get':
            case 'count':
                return this.count(res, req, undefined);
        }
        return this.getArray(res, req, undefined);
    };
    DataApi.prototype.httpPost = function (res, req, body, serializeContext) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var action, _a;
            return tslib_1.__generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        action = req === null || req === void 0 ? void 0 : req.get('__action');
                        if (action === null || action === void 0 ? void 0 : action.startsWith(rest_data_provider_js_1.liveQueryAction))
                            return [2 /*return*/, this.liveQuery(res, req, body, serializeContext, action.substring(rest_data_provider_js_1.liveQueryAction.length))];
                        _a = action;
                        switch (_a) {
                            case 'get': return [3 /*break*/, 1];
                            case 'count': return [3 /*break*/, 2];
                            case 'endLiveQuery': return [3 /*break*/, 3];
                        }
                        return [3 /*break*/, 5];
                    case 1: return [2 /*return*/, this.getArray(res, req, body)];
                    case 2: return [2 /*return*/, this.count(res, req, body)];
                    case 3: return [4 /*yield*/, this.remult.liveQueryStorage.remove(body.id)];
                    case 4:
                        _b.sent();
                        res.success('ok');
                        return [2 /*return*/];
                    case 5: return [2 /*return*/, this.post(res, body)];
                }
            });
        });
    };
    DataApi.prototype.get = function (response, id) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var _this = this;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!this.repository.metadata.apiReadAllowed) {
                            response.forbidden();
                            return [2 /*return*/];
                        }
                        return [4 /*yield*/, this.doOnId(response, id, function (row) { return tslib_1.__awaiter(_this, void 0, void 0, function () { return tslib_1.__generator(this, function (_a) {
                                return [2 /*return*/, response.success(this.repository.getEntityRef(row).toApiJson())];
                            }); }); })];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    DataApi.prototype.count = function (response, request, filterBody) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var _a, _b, _c, _d, err_1;
            var _e;
            return tslib_1.__generator(this, function (_f) {
                switch (_f.label) {
                    case 0:
                        if (!this.repository.metadata.apiReadAllowed) {
                            response.forbidden();
                            return [2 /*return*/];
                        }
                        _f.label = 1;
                    case 1:
                        _f.trys.push([1, 4, , 5]);
                        _b = (_a = response).success;
                        _e = {};
                        _d = (_c = this.repository).count;
                        return [4 /*yield*/, this.buildWhere(request, filterBody)];
                    case 2: return [4 /*yield*/, _d.apply(_c, [_f.sent()])];
                    case 3:
                        _b.apply(_a, [(_e.count = +(_f.sent()),
                                _e)]);
                        return [3 /*break*/, 5];
                    case 4:
                        err_1 = _f.sent();
                        response.error(err_1);
                        return [3 /*break*/, 5];
                    case 5: return [2 /*return*/];
                }
            });
        });
    };
    DataApi.prototype.getArrayImpl = function (response, request, filterBody) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var findOptions, _a, sort, dir, limit, hasId_1, w, r;
            var _this = this;
            return tslib_1.__generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        findOptions = {
                            load: function () { return []; },
                            include: this.includeNone(),
                        };
                        _a = findOptions;
                        return [4 /*yield*/, this.buildWhere(request, filterBody)];
                    case 1:
                        _a.where = _b.sent();
                        if (request) {
                            sort = request.get('_sort');
                            if (sort != undefined) {
                                dir = request.get('_order');
                                findOptions.orderBy = determineSort(sort, dir);
                            }
                            limit = +request.get('_limit');
                            if (!limit && DataApi.defaultGetLimit)
                                limit = DataApi.defaultGetLimit;
                            findOptions.limit = limit;
                            findOptions.page = +request.get('_page');
                        }
                        if (!this.remult.isAllowed(this.repository.metadata.options.apiRequireId)) return [3 /*break*/, 3];
                        hasId_1 = false;
                        return [4 /*yield*/, filter_interfaces_js_1.Filter.fromEntityFilter(this.repository.metadata, findOptions.where)];
                    case 2:
                        w = _b.sent();
                        if (w) {
                            w.__applyToConsumer({
                                containsCaseInsensitive: function () { },
                                notContainsCaseInsensitive: function () { },
                                isDifferentFrom: function () { },
                                isEqualTo: function (col, val) {
                                    if (_this.repository.metadata.idMetadata.isIdField(col))
                                        hasId_1 = true;
                                },
                                custom: function () { },
                                databaseCustom: function () { },
                                isGreaterOrEqualTo: function () { },
                                isGreaterThan: function () { },
                                isIn: function (col) {
                                    if (_this.repository.metadata.idMetadata.isIdField(col))
                                        hasId_1 = true;
                                },
                                isLessOrEqualTo: function () { },
                                isLessThan: function () { },
                                isNotNull: function () { },
                                isNull: function () { },
                                or: function () { },
                            });
                        }
                        if (!hasId_1) {
                            response.forbidden();
                            throw new server_action_js_1.ForbiddenError();
                        }
                        _b.label = 3;
                    case 3: return [4 /*yield*/, this.repository.find(findOptions).then(function (r) { return tslib_1.__awaiter(_this, void 0, void 0, function () {
                            var _this = this;
                            return tslib_1.__generator(this, function (_a) {
                                switch (_a.label) {
                                    case 0: return [4 /*yield*/, Promise.all(r.map(function (y) { return tslib_1.__awaiter(_this, void 0, void 0, function () { return tslib_1.__generator(this, function (_a) {
                                            return [2 /*return*/, this.repository.getEntityRef(y).toApiJson()];
                                        }); }); }))];
                                    case 1: return [2 /*return*/, _a.sent()];
                                }
                            });
                        }); })];
                    case 4:
                        r = _b.sent();
                        return [2 /*return*/, { r: r, findOptions: findOptions }];
                }
            });
        });
    };
    DataApi.prototype.includeNone = function () {
        var e_1, _a;
        var include = {};
        try {
            for (var _b = tslib_1.__values(this.repository.metadata.fields), _c = _b.next(); !_c.done; _c = _b.next()) {
                var field = _c.value;
                if ((0, relationInfoMember_js_1.getRelationFieldInfo)(field)) {
                    include[field.key] = false;
                }
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_1) throw e_1.error; }
        }
        return include;
    };
    DataApi.prototype.getArray = function (response, request, filterBody) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var r, err_2;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!this.repository.metadata.apiReadAllowed) {
                            response.forbidden();
                            return [2 /*return*/];
                        }
                        _a.label = 1;
                    case 1:
                        _a.trys.push([1, 3, , 4]);
                        return [4 /*yield*/, this.getArrayImpl(response, request, filterBody)];
                    case 2:
                        r = (_a.sent()).r;
                        response.success(r);
                        return [3 /*break*/, 4];
                    case 3:
                        err_2 = _a.sent();
                        if (err_2.isForbiddenError)
                            response.forbidden();
                        else
                            response.error(err_2);
                        return [3 /*break*/, 4];
                    case 4: return [2 /*return*/];
                }
            });
        });
    };
    DataApi.prototype.liveQuery = function (response, request, filterBody, serializeContext, queryChannel) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var r, data, err_3;
            var _a;
            var _this = this;
            return tslib_1.__generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        if (!this.repository.metadata.apiReadAllowed) {
                            response.forbidden();
                            return [2 /*return*/];
                        }
                        _b.label = 1;
                    case 1:
                        _b.trys.push([1, 5, , 6]);
                        return [4 /*yield*/, this.getArrayImpl(response, request, filterBody)];
                    case 2:
                        r = _b.sent();
                        _a = {};
                        return [4 /*yield*/, serializeContext()];
                    case 3:
                        data = (_a.requestJson = _b.sent(),
                            _a.findOptionsJson = (0, rest_data_provider_js_1.findOptionsToJson)(r.findOptions, this.repository.metadata),
                            _a.lastIds = r.r.map(function (y) { return _this.repository.metadata.idMetadata.getId(y); }),
                            _a);
                        return [4 /*yield*/, this.remult.liveQueryStorage.add({
                                entityKey: this.repository.metadata.key,
                                id: queryChannel,
                                data: data,
                            })];
                    case 4:
                        _b.sent();
                        response.success(r.r);
                        return [3 /*break*/, 6];
                    case 5:
                        err_3 = _b.sent();
                        if (err_3.isForbiddenError)
                            response.forbidden();
                        else
                            response.error(err_3);
                        return [3 /*break*/, 6];
                    case 6: return [2 /*return*/];
                }
            });
        });
    };
    DataApi.prototype.buildWhere = function (request, filterBody) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var where, _a, _b;
            return tslib_1.__generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        where = [];
                        if (!this.repository.metadata.options.apiPrefilter) return [3 /*break*/, 3];
                        if (!(typeof this.repository.metadata.options.apiPrefilter === 'function')) return [3 /*break*/, 2];
                        _b = (_a = where).push;
                        return [4 /*yield*/, this.repository.metadata.options.apiPrefilter()];
                    case 1:
                        _b.apply(_a, [_c.sent()]);
                        return [3 /*break*/, 3];
                    case 2:
                        where.push(this.repository.metadata.options.apiPrefilter);
                        _c.label = 3;
                    case 3:
                        if (request) {
                            where.push((0, filter_interfaces_js_1.buildFilterFromRequestParameters)(this.repository.metadata, {
                                get: function (key) {
                                    var result = request.get(key);
                                    if (key.startsWith(filter_interfaces_js_1.customUrlToken) &&
                                        result &&
                                        typeof result === 'string')
                                        return JSON.parse(result);
                                    return result;
                                },
                            }));
                        }
                        if (filterBody)
                            where.push(filter_interfaces_js_1.Filter.entityFilterFromJson(this.repository.metadata, filterBody));
                        return [2 /*return*/, { $and: where }];
                }
            });
        });
    };
    DataApi.prototype.doOnId = function (response, id, what) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var where, _a, _b, err_4;
            var _this = this;
            return tslib_1.__generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        _c.trys.push([0, 5, , 6]);
                        where = [
                            this.repository.metadata.idMetadata.getIdFilter(id),
                        ];
                        if (!this.repository.metadata.options.apiPrefilter) return [3 /*break*/, 3];
                        if (!(typeof this.repository.metadata.options.apiPrefilter === 'function')) return [3 /*break*/, 2];
                        _b = (_a = where).push;
                        return [4 /*yield*/, this.repository.metadata.options.apiPrefilter()];
                    case 1:
                        _b.apply(_a, [_c.sent()]);
                        return [3 /*break*/, 3];
                    case 2:
                        where.push(this.repository.metadata.options.apiPrefilter);
                        _c.label = 3;
                    case 3: return [4 /*yield*/, this.repository
                            .find({
                            where: { $and: where },
                            include: this.includeNone(),
                        })
                            .then(function (r) { return tslib_1.__awaiter(_this, void 0, void 0, function () {
                            return tslib_1.__generator(this, function (_a) {
                                switch (_a.label) {
                                    case 0:
                                        if (!(r.length == 0)) return [3 /*break*/, 1];
                                        response.notFound();
                                        return [3 /*break*/, 4];
                                    case 1:
                                        if (!(r.length > 1)) return [3 /*break*/, 2];
                                        response.error({ message: 'id is not unique' });
                                        return [3 /*break*/, 4];
                                    case 2: return [4 /*yield*/, what(r[0])];
                                    case 3:
                                        _a.sent();
                                        _a.label = 4;
                                    case 4: return [2 /*return*/];
                                }
                            });
                        }); })];
                    case 4:
                        _c.sent();
                        return [3 /*break*/, 6];
                    case 5:
                        err_4 = _c.sent();
                        response.error(err_4);
                        return [3 /*break*/, 6];
                    case 6: return [2 /*return*/];
                }
            });
        });
    };
    DataApi.prototype.put = function (response, id, body) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var _this = this;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.doOnId(response, id, function (row) { return tslib_1.__awaiter(_this, void 0, void 0, function () {
                            var ref;
                            return tslib_1.__generator(this, function (_a) {
                                switch (_a.label) {
                                    case 0:
                                        ref = this.repository.getEntityRef(row);
                                        return [4 /*yield*/, ref._updateEntityBasedOnApi(body)];
                                    case 1:
                                        _a.sent();
                                        if (!ref.apiUpdateAllowed) {
                                            response.forbidden();
                                            return [2 /*return*/];
                                        }
                                        return [4 /*yield*/, ref.save()];
                                    case 2:
                                        _a.sent();
                                        response.success(ref.toApiJson());
                                        return [2 /*return*/];
                                }
                            });
                        }); })];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    DataApi.prototype.delete = function (response, id) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var _this = this;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.doOnId(response, id, function (row) { return tslib_1.__awaiter(_this, void 0, void 0, function () {
                            return tslib_1.__generator(this, function (_a) {
                                switch (_a.label) {
                                    case 0:
                                        if (!this.repository.getEntityRef(row).apiDeleteAllowed) {
                                            response.forbidden();
                                            return [2 /*return*/];
                                        }
                                        return [4 /*yield*/, this.repository.getEntityRef(row).delete()];
                                    case 1:
                                        _a.sent();
                                        response.deleted();
                                        return [2 /*return*/];
                                }
                            });
                        }); })];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    DataApi.prototype.post = function (response, body) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var insert_1, result_1, _a, _b, err_5;
            var _this = this;
            return tslib_1.__generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        _c.trys.push([0, 5, , 6]);
                        insert_1 = function (what) { return tslib_1.__awaiter(_this, void 0, void 0, function () {
                            var newr;
                            return tslib_1.__generator(this, function (_a) {
                                switch (_a.label) {
                                    case 0:
                                        newr = this.repository.create();
                                        return [4 /*yield*/, this.repository.getEntityRef(newr)._updateEntityBasedOnApi(what)];
                                    case 1:
                                        _a.sent();
                                        if (!this.repository.getEntityRef(newr).apiInsertAllowed) {
                                            throw new server_action_js_1.ForbiddenError();
                                        }
                                        return [4 /*yield*/, this.repository.getEntityRef(newr).save()];
                                    case 2:
                                        _a.sent();
                                        return [2 /*return*/, this.repository.getEntityRef(newr).toApiJson()];
                                }
                            });
                        }); };
                        if (!Array.isArray(body)) return [3 /*break*/, 2];
                        result_1 = [];
                        return [4 /*yield*/, (0, context_js_1.doTransaction)(this.remult, function () { return tslib_1.__awaiter(_this, void 0, void 0, function () {
                                var body_1, body_1_1, item, _a, _b, e_2_1;
                                var e_2, _c;
                                return tslib_1.__generator(this, function (_d) {
                                    switch (_d.label) {
                                        case 0:
                                            _d.trys.push([0, 5, 6, 7]);
                                            body_1 = tslib_1.__values(body), body_1_1 = body_1.next();
                                            _d.label = 1;
                                        case 1:
                                            if (!!body_1_1.done) return [3 /*break*/, 4];
                                            item = body_1_1.value;
                                            _b = (_a = result_1).push;
                                            return [4 /*yield*/, insert_1(item)];
                                        case 2:
                                            _b.apply(_a, [_d.sent()]);
                                            _d.label = 3;
                                        case 3:
                                            body_1_1 = body_1.next();
                                            return [3 /*break*/, 1];
                                        case 4: return [3 /*break*/, 7];
                                        case 5:
                                            e_2_1 = _d.sent();
                                            e_2 = { error: e_2_1 };
                                            return [3 /*break*/, 7];
                                        case 6:
                                            try {
                                                if (body_1_1 && !body_1_1.done && (_c = body_1.return)) _c.call(body_1);
                                            }
                                            finally { if (e_2) throw e_2.error; }
                                            return [7 /*endfinally*/];
                                        case 7: return [2 /*return*/];
                                    }
                                });
                            }); })];
                    case 1:
                        _c.sent();
                        response.created(result_1);
                        return [3 /*break*/, 4];
                    case 2:
                        _b = (_a = response).created;
                        return [4 /*yield*/, insert_1(body)];
                    case 3:
                        _b.apply(_a, [_c.sent()]);
                        _c.label = 4;
                    case 4: return [3 /*break*/, 6];
                    case 5:
                        err_5 = _c.sent();
                        if (err_5.isForbiddenError)
                            response.forbidden();
                        else
                            response.error(err_5);
                        return [3 /*break*/, 6];
                    case 6: return [2 /*return*/];
                }
            });
        });
    };
    DataApi.defaultGetLimit = 0;
    return DataApi;
}());
exports.DataApi = DataApi;
function determineSort(sortUrlParm, dirUrlParam) {
    var dirItems = [];
    if (dirUrlParam)
        dirItems = dirUrlParam.split(',');
    var result = {};
    sortUrlParm.split(',').map(function (name, i) {
        var key = name.trim();
        if (i < dirItems.length && dirItems[i].toLowerCase().trim().startsWith('d'))
            return (result[key] = 'desc');
        else
            return (result[key] = 'asc');
    });
    return result;
}
exports.determineSort = determineSort;
function serializeError(data) {
    if (data instanceof TypeError) {
        data = { message: data.message, stack: data.stack };
    }
    var x = JSON.parse(JSON.stringify(data));
    if (!x.message && !x.modelState)
        data = { message: data.message, stack: data.stack };
    if (typeof x === 'string')
        data = { message: x };
    return data;
}
exports.serializeError = serializeError;
