// remult/src/remult-static.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.defaultFactory = exports.remultStatic = void 0;
var remultStaticKey = Symbol.for('remult-static1');
var x = {
    defaultRemultFactory: undefined,
    defaultRemult: undefined,
    asyncContext: undefined,
    columnsOfType: new Map(),
    allEntities: [],
    classHelpers: new Map(),
    actionInfo: {
        allActions: [],
        runningOnServer: false,
        runActionWithoutBlockingUI: function (what) {
            return what();
        },
        startBusyWithProgress: function () { return ({
            progress: function (percent) { },
            close: function () { },
        }); },
    },
    captionTransformer: undefined,
};
if ((typeof process !== 'undefined' &&
    process.env['IGNORE_GLOBAL_REMULT_IN_TESTS']) ||
    typeof globalThis[remultStaticKey] === 'undefined') {
    globalThis[remultStaticKey] = x;
}
else {
    x = globalThis[remultStaticKey];
}
exports.remultStatic = x;
function defaultFactory() {
    if (!exports.remultStatic.defaultRemult) {
        exports.remultStatic.defaultRemult = exports.remultStatic.defaultRemultFactory();
    }
    return exports.remultStatic.defaultRemult;
}
exports.defaultFactory = defaultFactory;
