// remult/src/sort.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Sort = void 0;
var tslib_1 = require("tslib");
var CompoundIdField_js_1 = require("./CompoundIdField.js");
var relationInfoMember_js_1 = require("./remult3/relationInfoMember.js");
var Sort = /** @class */ (function () {
    function Sort() {
        var segments = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            segments[_i] = arguments[_i];
        }
        this.Segments = segments;
    }
    Sort.prototype.toEntityOrderBy = function () {
        var e_1, _a;
        var result = {};
        try {
            for (var _b = tslib_1.__values(this.Segments), _c = _b.next(); !_c.done; _c = _b.next()) {
                var seg = _c.value;
                if (seg.isDescending) {
                    result[seg.field.key] = 'desc';
                }
                else
                    result[seg.field.key] = 'asc';
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_1) throw e_1.error; }
        }
        return result;
    };
    Sort.prototype.reverse = function () {
        var e_2, _a;
        var r = new Sort();
        try {
            for (var _b = tslib_1.__values(this.Segments), _c = _b.next(); !_c.done; _c = _b.next()) {
                var s = _c.value;
                r.Segments.push({ field: s.field, isDescending: !s.isDescending });
            }
        }
        catch (e_2_1) { e_2 = { error: e_2_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_2) throw e_2.error; }
        }
        return r;
    };
    Sort.prototype.compare = function (a, b, getFieldKey) {
        if (!getFieldKey)
            getFieldKey = function (x) { return x.key; };
        var r = 0;
        for (var i = 0; i < this.Segments.length; i++) {
            var seg = this.Segments[i];
            var left = fixValueForSort(a[getFieldKey(seg.field)]);
            var right = fixValueForSort(b[getFieldKey(seg.field)]);
            if (left > right)
                r = 1;
            else if (left < right)
                r = -1;
            if (r != 0) {
                if (seg.isDescending)
                    r *= -1;
                return r;
            }
        }
        return r;
    };
    Sort.translateOrderByToSort = function (entityDefs, orderBy) {
        if (!orderBy)
            return undefined;
        var sort = new Sort();
        if (orderBy) {
            var _loop_1 = function (key) {
                if (Object.prototype.hasOwnProperty.call(orderBy, key)) {
                    var element_1 = orderBy[key];
                    var field = entityDefs.fields.find(key);
                    var addSegment = function (field) {
                        switch (element_1) {
                            case 'desc':
                                sort.Segments.push({ field: field, isDescending: true });
                                break;
                            case 'asc':
                                sort.Segments.push({ field: field });
                        }
                    };
                    if (field) {
                        var rel = (0, relationInfoMember_js_1.getRelationFieldInfo)(field);
                        if ((rel === null || rel === void 0 ? void 0 : rel.type) === 'toOne') {
                            var op = rel.options;
                            if (typeof op.field === 'string') {
                                addSegment(entityDefs.fields.find(op.field));
                            }
                            else {
                                if (op.fields) {
                                    for (var key_1 in op.fields) {
                                        if (Object.prototype.hasOwnProperty.call(op.fields, key_1)) {
                                            var keyInMyEntity = op.fields[key_1];
                                            addSegment(entityDefs.fields.find(keyInMyEntity.toString()));
                                        }
                                    }
                                }
                            }
                        }
                        else
                            addSegment(field);
                    }
                }
            };
            for (var key in orderBy) {
                _loop_1(key);
            }
        }
        return sort;
    };
    Sort.createUniqueSort = function (entityMetadata, orderBy) {
        var e_3, _a;
        if ((!orderBy || Object.keys(orderBy).length === 0) &&
            entityMetadata.options.defaultOrderBy)
            orderBy = Sort.translateOrderByToSort(entityMetadata, entityMetadata.options.defaultOrderBy);
        if (!orderBy)
            orderBy = new Sort({ field: entityMetadata.idMetadata.field });
        if (entityMetadata.idMetadata.field instanceof CompoundIdField_js_1.CompoundIdField) {
            var _loop_2 = function (field) {
                if (!orderBy.Segments.find(function (x) { return x.field == field; })) {
                    orderBy.Segments.push({ field: field });
                }
            };
            try {
                for (var _b = tslib_1.__values(entityMetadata.idMetadata.field.fields), _c = _b.next(); !_c.done; _c = _b.next()) {
                    var field = _c.value;
                    _loop_2(field);
                }
            }
            catch (e_3_1) { e_3 = { error: e_3_1 }; }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                }
                finally { if (e_3) throw e_3.error; }
            }
        }
        else if (!orderBy.Segments.find(function (x) { return x.field == entityMetadata.idMetadata.field; })) {
            orderBy.Segments.push({ field: entityMetadata.idMetadata.field });
        }
        return orderBy;
    };
    Sort.createUniqueEntityOrderBy = function (entityMetadata, orderBy) {
        var _a, e_4, _b;
        if (!orderBy || Object.keys(orderBy).length === 0)
            orderBy = entityMetadata.options.defaultOrderBy;
        if (!orderBy)
            orderBy = (_a = {},
                _a[entityMetadata.idMetadata.field.key] = 'asc',
                _a);
        else
            orderBy = tslib_1.__assign({}, orderBy);
        if (entityMetadata.idMetadata.field instanceof CompoundIdField_js_1.CompoundIdField) {
            try {
                for (var _c = tslib_1.__values(entityMetadata.idMetadata.field.fields), _d = _c.next(); !_d.done; _d = _c.next()) {
                    var field = _d.value;
                    if (!orderBy[field.key]) {
                        orderBy[field.key] = 'asc';
                    }
                }
            }
            catch (e_4_1) { e_4 = { error: e_4_1 }; }
            finally {
                try {
                    if (_d && !_d.done && (_b = _c.return)) _b.call(_c);
                }
                finally { if (e_4) throw e_4.error; }
            }
        }
        else if (!orderBy[entityMetadata.idMetadata.field.key]) {
            orderBy[entityMetadata.idMetadata.field.key] = 'asc';
        }
        return orderBy;
    };
    return Sort;
}());
exports.Sort = Sort;
function fixValueForSort(a) {
    if (a == undefined || a == null)
        return a;
    if (a.id !== undefined)
        return a.id;
    return a;
}
